File: python/ch13/loop_toy.py
# python/ch13/loop_toy.py
"""AHC010 簡略: ループ用辺を貪欲配置し UF で連結性を維持."""
from __future__ import annotations

import sys
import time
from pathlib import Path
from typing import List, Set, Tuple

from union_find import UnionFind


def read_grid(stream) -> Tuple[int, int, List[str]]:
    lines = stream.read().splitlines()
    h, w = map(int, lines[0].split())
    grid = [lines[i + 1] for i in range(h)]
    return h, w, grid


def cell_index(r: int, c: int, w: int) -> int:
    return r * w + c


def neighbors(r: int, c: int, h: int, w: int) -> List[Tuple[int, int]]:
    out = []
    if r > 0:
        out.append((r - 1, c))
    if r + 1 < h:
        out.append((r + 1, c))
    if c > 0:
        out.append((r, c - 1))
    if c + 1 < w:
        out.append((r, c + 1))
    return out


def full_recompute_components(h: int, w: int, edges: Set[Tuple[int, int]]) -> int:
    n = h * w
    parent = list(range(n))

    def find(x: int) -> int:
        while parent[x] != x:
            x = parent[x]
        return x

    for u, v in edges:
        ru, rv = find(u), find(v)
        if ru != rv:
            parent[rv] = ru
    return len({find(i) for i in range(n)})


def greedy_loop_edges(h: int, w: int, grid: List[str], use_uf: bool) -> Tuple[int, Set[Tuple[int, int]]]:
    edges: Set[Tuple[int, int]] = set()
    uf = UnionFind(h * w) if use_uf else None
    placed = 0
    for r in range(h):
        for c in range(w):
            if grid[r][c] == "#":
                continue
            here = cell_index(r, c, w)
            for nr, nc in neighbors(r, c, h, w):
                if grid[nr][nc] == "#":
                    continue
                if (nr, nc) < (r, c):
                    continue
                there = cell_index(nr, nc, w)
                key = (min(here, there), max(here, there))
                if use_uf:
                    if uf.unite(here, there):
                        edges.add(key)
                        placed += 1
                else:
                    trial = set(edges)
                    trial.add(key)
                    comps = full_recompute_components(h, w, trial)
                    if comps <= h * w - len(trial) + 1:
                        edges.add(key)
                        placed += 1
    return placed, edges


def benchmark_connectivity(h: int, w: int, grid: List[str], trials: int = 200) -> None:
    t0 = time.perf_counter()
    for _ in range(trials):
        greedy_loop_edges(h, w, grid, use_uf=False)
    naive_ms = (time.perf_counter() - t0) * 1000

    t0 = time.perf_counter()
    for _ in range(trials):
        greedy_loop_edges(h, w, grid, use_uf=True)
    uf_ms = (time.perf_counter() - t0) * 1000

    print(f"connectivity_trials={trials}")
    print(f"full_recompute_ms={naive_ms:.2f}")
    print(f"union_find_ms={uf_ms:.2f}")
    print(f"speedup={naive_ms / max(uf_ms, 1e-9):.1f}x")


def main() -> None:
    if len(sys.argv) > 1:
        path = Path(sys.argv[1])
        text = path.read_text()
    else:
        text = sys.stdin.read()
    from io import StringIO

    h, w, grid = read_grid(StringIO(text))
    placed, edges = greedy_loop_edges(h, w, grid, use_uf=True)
    comps = full_recompute_components(h, w, edges)
    print(f"grid={h}x{w} edges={len(edges)} placed={placed} components={comps}")
    benchmark_connectivity(h, w, grid, trials=80)


if __name__ == "__main__":
    main()

File: go/ch13/loop_toy.go
// go/ch13/loop_toy.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type loopUF struct {
	parent []int
	rank   []int
}

func newLoopUF(n int) *loopUF {
	p := make([]int, n)
	for i := range p {
		p[i] = i
	}
	return &loopUF{parent: p, rank: make([]int, n)}
}

func (uf *loopUF) Find(x int) int {
	for uf.parent[x] != x {
		uf.parent[x] = uf.parent[uf.parent[x]]
		x = uf.parent[x]
	}
	return x
}

func (uf *loopUF) Unite(a, b int) bool {
	ra, rb := uf.Find(a), uf.Find(b)
	if ra == rb {
		return false
	}
	if uf.rank[ra] < uf.rank[rb] {
		ra, rb = rb, ra
	}
	uf.parent[rb] = ra
	if uf.rank[ra] == uf.rank[rb] {
		uf.rank[ra]++
	}
	return true
}

func readGrid(sc *bufio.Scanner) (int, int, []string, error) {
	sc.Scan()
	p := strings.Fields(sc.Text())
	h, _ := strconv.Atoi(p[0])
	w, _ := strconv.Atoi(p[1])
	grid := make([]string, h)
	for i := 0; i < h; i++ {
		sc.Scan()
		grid[i] = sc.Text()
	}
	return h, w, grid, sc.Err()
}

func cellIndex(r, c, w int) int {
	return r*w + c
}

func gridNeighbors(r, c, h, w int) [][2]int {
	var out [][2]int
	if r > 0 {
		out = append(out, [2]int{r - 1, c})
	}
	if r+1 < h {
		out = append(out, [2]int{r + 1, c})
	}
	if c > 0 {
		out = append(out, [2]int{r, c - 1})
	}
	if c+1 < w {
		out = append(out, [2]int{r, c + 1})
	}
	return out
}

func fullRecompute(h, w int, edges map[[2]int]bool) int {
	n := h * w
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	find := func(x int) int {
		for parent[x] != x {
			x = parent[x]
		}
		return x
	}
	for k := range edges {
		ru, rv := find(k[0]), find(k[1])
		if ru != rv {
			parent[rv] = ru
		}
	}
	roots := make(map[int]bool)
	for i := 0; i < n; i++ {
		roots[find(i)] = true
	}
	return len(roots)
}

func greedyLoop(h, w int, grid []string, useUF bool) (int, map[[2]int]bool) {
	edges := make(map[[2]int]bool)
	uf := newLoopUF(h * w)
	placed := 0
	for r := 0; r < h; r++ {
		for c := 0; c < w; c++ {
			if grid[r][c] == '#' {
				continue
			}
			here := cellIndex(r, c, w)
			for _, nb := range gridNeighbors(r, c, h, w) {
				nr, nc := nb[0], nb[1]
				if grid[nr][nc] == '#' {
					continue
				}
				if nr < r || (nr == r && nc < c) {
					continue
				}
				there := cellIndex(nr, nc, w)
				key := [2]int{here, there}
				if here > there {
					key = [2]int{there, here}
				}
				if useUF {
					if uf.Unite(here, there) {
						edges[key] = true
						placed++
					}
				} else {
					trial := make(map[[2]int]bool)
					for k := range edges {
						trial[k] = true
					}
					trial[key] = true
					comps := fullRecompute(h, w, trial)
					if comps <= h*w-len(trial)+1 {
						edges[key] = true
						placed++
					}
				}
			}
		}
	}
	return placed, edges
}

func benchmarkConnectivity(h, w int, grid []string, trials int) {
	start := time.Now()
	for t := 0; t < trials; t++ {
		greedyLoop(h, w, grid, false)
	}
	naiveMs := time.Since(start).Seconds() * 1000

	start = time.Now()
	for t := 0; t < trials; t++ {
		greedyLoop(h, w, grid, true)
	}
	ufMs := time.Since(start).Seconds() * 1000

	fmt.Printf("connectivity_trials=%d\n", trials)
	fmt.Printf("full_recompute_ms=%.2f\n", naiveMs)
	fmt.Printf("union_find_ms=%.2f\n", ufMs)
	fmt.Printf("speedup=%.1fx\n", naiveMs/max(ufMs, 1e-9))
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	h, w, grid, err := readGrid(sc)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	placed, edges := greedyLoop(h, w, grid, true)
	comps := fullRecompute(h, w, edges)
	fmt.Printf("grid=%dx%d edges=%d placed=%d components=%d\n", h, w, len(edges), placed, comps)
	benchmarkConnectivity(h, w, grid, 80)
}
